packages/monaco/src/browser/monaco-workspace.ts

```diff
@@ -170,13 +170,24 @@ export class MonacoWorkspace implements lang.Workspace {
         const timeout = new Promise<TextEdit[]>(resolve =>
             setTimeout(() => resolve([]), 1000)
         );
-        const resolveEdits = new Promise<TextEdit[]>(resolve =>
+        const resolveEdits = new Promise<TextEdit[]>(async resolve => {
+            const thenables: Thenable<TextEdit[]>[] = [];
+            const allEdits: TextEdit[] = [];
+
             this.onWillSaveTextDocumentEmitter.fire({
                 textDocument: event.model,
                 reason,
-                waitUntil: thenable => thenable.then(resolve)
-            })
-        );
+                waitUntil: thenable => {
+                    thenables.push(thenable);
+                }
+            });
+
+            for (const listenerEdits of await Promise.all(thenables)) {
+                allEdits.push(...listenerEdits);
```

Comment:
It's still a mystery to me how edits from multiple sources, based on the same initial version, can apply cleanly in succession.  But if it works for our current use case, I won't ask questions.

Reply:
In this case there is no listeners (which was also the reason why it never resolved before), so no collisions either :p

But I understand your concern, no idea how it works, maybe it is bogus, but I would like to keep that for later as it is beyond my understanding now.

Reply:
@svenefftinge do you or someone else know anything about this API and how it works?

Reply:
IIUC, in VS Code they do contribute [in sequence, no parallelism](https://github.com/Microsoft/vscode/blob/master/src/vs/workbench/api/node/extHostDocumentSaveParticipant.ts#L58-L69).

Reply:
So I don't really know how to mimic the behavior in VS Code without changing the API in Theia...

Reply:
> So I don't really know how to mimic the behavior in VS Code without changing the API in Theia...

Do you mean the first contributor works with the original version of the document, then second contributor works with the version already modified by the first one, etc?

Reply:
> In this case there is no listeners (which was also the reason why it never resolved before), so no collisions either :p
> 
> But I understand your concern, no idea how it works, maybe it is bogus, but I would like to keep that for later as it is beyond my understanding now.

Yes, I think we can merge your patch, but keep this in mind for later.  We can maybe come up with a synthetic test with two or more contributors that try to edit the same line to see what happens.

Reply:
> Do you mean the first contributor works with the original version of the document, then second contributor works with the version already modified by the first one, etc?

Yes.